libcxx/test/std/utilities/memory/pointer.conversion/to_address_without_pointer_traits.pass.cpp

```diff
@@ -0,0 +1,48 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <memory>
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+
+// template <class Ptr> constexpr auto to_address(const Ptr& p) noexcept;
+//     Should not require a specialization of pointer_traits for Ptr.
+
+#include <memory>
+
+struct IntPtr {
+  constexpr int* operator->() const { return ptr; }
+
+  int* ptr;
+};
+
```

Comment:
You need to add some tests for the actual SFINAE friendliness of the type. You'll need some sort of metafunction that allows you to test code in a SFINAE context. For example:

```
template <template <class...> class Templ, class Ignore, class... Args>
struct is_valid_expansion_impl
{
    typedef std::false_type type;
};

template <template <class...> class Templ, class... Args>
struct is_valid_expansion_impl<Templ, decltype((void) Templ<Args...>{}, 0), Args...>
{
    typedef std::true_type type;
};

template <template <class...> class Templ, class... Args>
using is_valid_expansion = typename is_valid_expansion_impl<Templ, int, Args...>::type;


template <class Ptr>
using TestToAddressCall = decltype(std::to_address(std::declval<Ptr>()))

static_assert(is_valid_expansion<TestToAddress, IntPtr>>::value);
static_assert(not is_valid_expansion<TestToAddress, std::string>>::value); 
```


At least that's the least ugly way I can think to test to_address without using concepts

Reply:
OK, that's very clever. I added the tests you suggested, as well as some direct tests of pointer_traits themselves.

It seems like rebind and pointer_to aren't SFINAE-friendly either, but I was able to test that pointer_traits for a non-pointer-like type is empty at least.

Reply:
By the way, do we generally follow LLVM naming conventions in libc++ tests? I'm not sure if it's better to name traits `like_this` or `LikeThat`.

Reply:
I think the snake_case you have right now is reasonable.